### PR TITLE
add test for fields of references

### PR DESCRIPTION
**Pull request by aneksteind**

### CODE REVIEW

File: pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap
Comment (spernsteiner) [inline pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap:700]: This has source `n[0]`, but I think it should be `n[2]`, since that's the last node that assigned to `_6`.

Comment (aneksteind) [reply pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap]: Fixed in 4ca53bd3

File: pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap
Comment (kkysen) [inline pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap:692]: What does `(Misc)` mean?

File: analysis/test/src/pointers.rs
Comment (kkysen) [inline analysis/test/src/pointers.rs:517]: Could you put the new `test_ref_field()` at the end so it doesn't disturb the snapshot as much?  Or is there a reason it's in that position?

File: pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap
Comment (kkysen) [inline pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap:168]: Does this have no source because it's a copy of a string literal?

File: pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap
Comment (kkysen) [reply pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap:692]: And in general the other things in the parentheses after the type.

File: dynamic_instrumentation/src/instrument.rs
Comment (kkysen) [inline dynamic_instrumentation/src/instrument.rs:245]: `remove_outer_deref(*p, self.tcx())` is already computed above on line 238:

```rust
                    && place_ty(&remove_outer_deref(*p, self.tcx())).is_region_ptr() =>
```

Could you de-duplicate this?

Comment (kkysen) [reply dynamic_instrumentation/src/instrument.rs:245]: Maybe a `try_remove_outer_deref` would be useful here:

```rust
/// Try to strip the initital [`Deref`](ProjectionElem::Deref)
/// from a [`projection`](PlaceRef::projection) sequence.
pub fn try_remove_outer_deref<'tcx>(p: Place<'tcx>, tcx: TyCtxt<'tcx>) -> Option<Place<'tcx>> {
    // Remove outer deref if present
    match p.as_ref() {
        PlaceRef {
            local,
            projection: &[ref base @ .., ProjectionElem::Deref],
        } => Some(Place {
            local,
            projection: tcx.intern_place_elems(base),
        }),
        _ => None,
    }
}

/// Strip the initital [`Deref`](ProjectionElem::Deref)
/// from a [`projection`](PlaceRef::projection) sequence
/// if there is one.
pub fn remove_outer_deref<'tcx>(p: Place<'tcx>, tcx: TyCtxt<'tcx>) -> Place<'tcx> {
    try_remove_outer_deref(p, tcx).unwrap_or(p)
}
```

Comment (kkysen) [reply dynamic_instrumentation/src/instrument.rs:245]: Could do

```rust
            Rvalue::AddressOf(_, p)
                if let Some(source) = try_remove_outer_deref(*p, self.tcx())
                && place_ty(source).is_region_ptr() =>
            {
                let instrumentation_location = Location {
                    statement_index: location.statement_index + 1,
                    ..location
                };
                // Instrument which local's address is taken
                self.loc(instrumentation_location, copy_fn)
                    .arg_var(dest)
                    .source(source)
                    .dest(&dest)
                    .debug_mir(location)
                    .add_to(self);
            }
```

but that requires `#![feature(if_let_guard, let_chains)]`.  It might be too clunky without those, I think.

Comment (kkysen) [reply dynamic_instrumentation/src/instrument.rs:236]: How about the changes in https://github.com/immunant/c2rust/pull/573?

Comment (kkysen) [reply dynamic_instrumentation/src/instrument.rs:236]: I think https://github.com/immunant/c2rust/pull/573 is probably too big for the scope of this PR, but keep it in my mind for after merging this PR.

File: analysis/test/src/pointers.rs
Comment (kkysen) [reply analysis/test/src/pointers.rs:517]: If you want to, it's done in https://github.com/immunant/c2rust/pull/574.  The diff with `master` is cleaner.

**Review by kkysen — APPROVED** [review]
See my improvement PRs.

File: pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap
Comment (spernsteiner) [reply pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap]: That's the [`CastKind`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/mir/enum.CastKind.html).  There are a few kinds of pointer coercions that have specific `CastKind`s, and everything else is `CastKind::Misc`.

File: analysis/test/src/pointers.rs
Comment (fw-immunant) [inline analysis/test/src/pointers.rs:386]: This pattern (`free` on a ptr-casted ref) is surprising to me--I don't think the transpiler emits this sort of code (it would pass the original raw ptr to free), and neither idiomatic nor partially-refactored, transpiled rust seems likely to behave this way either (the latter would presumably not make `s_ref` a mutable reference if it saw it being passed to `free`). This doesn't mean this isn't a reasonable way to test accesses to fields of references for instrumentation, but it did raise some eyebrows. Are we avoiding using a stack `S` here because `AddrOfLocal` support is still WIP?

Comment (fw-immunant) [reply analysis/test/src/pointers.rs:386]: `calloc(0, ...)` here is not sane since we do read/write into the allocation.

**Review by fw-immunant — APPROVED** [review]
Approved conditional on fixing the `calloc(0, ...)` in the test case.

File: analysis/test/src/pointers.rs
Comment (kkysen) [reply analysis/test/src/pointers.rs:386]: It should be `1`, right?  How come that doesn't segfault on a null pointer then?

File: analysis/test/src/pointers.rs
Comment (kkysen) [inline analysis/test/src/pointers.rs:381]:
```suggestion
        1i32 as libc::c_ulong,
```

File: pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap
Comment (aneksteind) [reply pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap]: it's because the rvalue of the statement that assigns to `_31` isn't yet instrumented, i.e. `Operand::Constant`. I'll make an issue for this

Comment (aneksteind) [reply pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap]: https://github.com/immunant/c2rust/issues/577

**Pull request merged**
